File: examples/python/abstract_reader.py
#/usr/bin/env python
# -*- coding: utf-8 -*-

# Description
# -----------
# Universal format reader.
# Input can be PE, ELF or Mach-O

import lief
import sys

def print_sections(sections):
    print("== Sections ==")
    f_title = "|{:<30} | {:<18}| {:<18}| {:<18}| {:<9}|"
    f_value = "|{:<30} | 0x{:<16x}| 0x{:<16x}| 0x{:<16x}| {:<9.2f}|"
    print(f_title.format("Name", "File offset", "Size", "Virtual Address", "Entropy"))
    for section in sections:
        print(f_value.format(\
                section.name,\
                section.offset,\
                section.size,\
                section.virtual_address,\
                section.entropy))
    print("")

def print_symbols(symbols):

    print("== Symbols ==")

    f = "|{:<30} |"

    print(f.format("Name"))

    for symbol in symbols:
        print(f.format(symbol.name))
    print("")


def print_exported_functions(functions):

    print("== Exported functions ==")
    f = "|{:<30} |"
    print(f.format("Name"))
    for func in functions:
            print(f.format(func))
    print("")

def print_imported_functions(functions):

    print("== Imported functions ==")
    f = "|{:<30} |"
    print(f.format("Name"))
    for func in functions:
        print(f.format(func))
    print("")


def print_imported_libraries(libraries):

    print("== Imported Libraries ==")
    f = "|{:<30} |"
    print(f.format("Name"))
    for library in libraries:
        print(f.format(library))
    print("")



def read_binary(path):
    print("== Abstract Reader ==")

    binary = lief.parse(path).abstract
    header = binary.header
    print(header)

    sections      = binary.sections
    symbols       = binary.symbols
    func_exported = binary.exported_functions
    func_imported = binary.imported_functions
    libraries     = binary.libraries

    if len(sections) > 0:
        print_sections(sections)

    if len(symbols) > 0:
        print_symbols(symbols)

    if len(func_exported) > 0:
        print_exported_functions(func_exported)

    if len(func_imported) > 0:
        print_imported_functions(func_imported)

    if len(libraries) > 0:
        print_imported_libraries(libraries)



if __name__ == "__main__":

    if len(sys.argv) != 2:
        print("Usage:", sys.argv[0], "<binary>")
        sys.exit(1)

    read_binary(sys.argv[1])

File: examples/cpp/abstract_reader.cpp
/* Copyright 2017 - 2024 R. Thomas
 * Copyright 2017 - 2024 Quarkslab
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <iostream>
#include <memory>

#include <LIEF/Abstract.hpp>

int main(int argc, char **argv) {
  std::cout << "Abstract Reader" << '\n';
  if (argc != 2) {
    std::cerr << "Usage: " << argv[0] << " <binary>" << '\n';
    return -1;
  }

  std::unique_ptr<const LIEF::Binary> binary{LIEF::Parser::parse(argv[1])};

  std::cout << "== Header ==" << '\n';
  std::cout << binary->header() << '\n';

  std::cout << "== Sections ==" << '\n';
  for (const LIEF::Section& s : binary->sections()) {
    std::cout << s << '\n';
  }

  std::cout << "== Symbols ==" << '\n';
  for (const LIEF::Symbol& s : binary->symbols()) {
    std::cout << s << '\n';
  }

  std::cout << "== Exported functions ==" << '\n';
  for(const LIEF::Function& func : binary->exported_functions()) {
    std::cout << func << '\n';
  }

  std::cout << "== Imported functions ==" << '\n';
  for(const LIEF::Function& func : binary->imported_functions()) {
    std::cout << func << '\n';
  }

  std::cout << "== Imported Libraries ==" << '\n';
  for(const std::string& name : binary->imported_libraries()) {
    std::cout << name << '\n';
  }

  std::cout << "== Relocation ==" << '\n';
  for(const LIEF::Relocation& relocation : binary->relocations()) {
    std::cout << relocation << '\n';
  }

  return 0;


}
